sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -1029,12 +1040,35 @@ public double getProgress() {
   private Progress getProgress() {
     synchronized (splitLock) {
       if (currentTracker instanceof RestrictionTracker.HasProgress) {
-        return ((HasProgress) currentTracker).getProgress();
+        return scaleProgress(
+            ((HasProgress) currentTracker).getProgress(),
+            currentWindowIterator.previousIndex(),
+            currentElement.getWindows().size());
       }
     }
     return null;
   }
 
+  private Progress getProgressFromWindowObservingTruncate(double elementCompleted) {
+    synchronized (splitLock) {
+      if (currentWindowIterator != null) {
+        return scaleProgress(
+            Progress.from(elementCompleted, 1 - elementCompleted),
+            currentWindowIterator.previousIndex(),
+            currentElement.getWindows().size());
+      }
+    }
+    return null;
+  }
+
+  private Progress scaleProgress(Progress progress, int completedWindowIndex, int windowCount) {
```

Comment:
For the future:
```suggestion
  private Progress scaleProgress(Progress progress, int currentWindowIndex, int stopWindowIndex) {
```

Note that it is important that it isn't the number of completed windows but the current window index otherwise the math doesn't work out.